src/Authentication/Authenticators/Session.php

```diff
@@ -249,28 +285,85 @@ public function loggedIn(): bool
         if ($userId !== null) {
             $this->user = $this->provider->findById($userId);
 
-            return $this->user instanceof User;
+            $identities = $this->userIdentityModel->getIdentitiesByTypes(
+                $this->user->getAuthId(),
+                ['email_2fa', 'email_activate']
+            );
+
+            // If we will have more than one identity, we need to change the logic blow.
+            assert(
+                count($identities) < 2,
+                'More than one identity for actions. user_id: ' . $userId
+            );
+
+            // Having an action?
+            foreach ($identities as $identity) {
+                $actionClass = setting('Auth.actions')[$identity->name];
+
+                if ($actionClass) {
+                    $this->userState = self::STATE_PENDING;
+
+                    session()->set('auth_action', $actionClass);
+                    $this->pendingMessage = $identity->extra;
+
+                    return;
+                }
+            }
+
+            $this->userState = self::STATE_LOGGED_IN;
+
+            return;
         }
 
         // Check remember-me token.
         if (setting('Auth.sessionConfig')['allowRemembering']) {
-            return $this->checkRememberMe();
+            $this->checkRememberMe();
+
+            return;
         }
 
-        return false;
+        $this->userState = self::STATE_ANONYMOUS;
+    }
+
+    public function isPending(): bool
+    {
+        $this->checkUserState();
+
+        return $this->userState === self::STATE_PENDING;
+    }
+
+    public function isAnonymous(): bool
```

Comment:
Please add a quick docblock with why a user might be anonymous

Reply:
Since `isPending()` was necessary, I also created `isAnonymous()` by the way.
But don't we need it? We are not using it at the moment.

Reply:
I don't have any concerns with the method/state. Those are both fine. But adding a description of what the `anonymous` represents within this library would save a developer a bit of time trying to parse through everything and understand.

Reply:
Added.